rewriter: create {add,rm}_extra_files commands

ElementW opened the pull request:
Add ability to mutate a target's `extra_files` list through the rewriter.

The logic is copied from sources add/rm, but changes the `extra_files` kwarg instead of the sources positional argument.
Has additional logic to handle creating the `extra_files` list if it doesn't exist.

mensinda commented on mesonbuild/rewriter.py:
```diff
@@ -709,6 +709,87 @@ def find_node(src):
                 if root not in self.modified_nodes:
                     self.modified_nodes += [root]
 
+        if cmd['operation'] == 'extra_files_add':
```
This doesn't really matter, but for consistency, this should also be an `elif`.

ElementW replied:
Great catch, didn't notice when I copy-pasted the sources add/rm logic. Will fix.

mensinda reviewed (commented):
Thanks for adding this. The PR looks mostly good, but needs tests and a [release note snippet](https://github.com/mesonbuild/meson/tree/master/docs/markdown/snippets).

For the existing tests:

- https://github.com/mesonbuild/meson/blob/master/unittests/rewritetests.py
- https://github.com/mesonbuild/meson/tree/master/test%20cases/rewrite

ElementW commented:
> The PR looks mostly good, but needs tests and a [release note snippet](https://github.com/mesonbuild/meson/tree/master/docs/markdown/snippets).

I did mention that:

> Unit test, doc and release note snippets will come, until then this is a draft PR.

mensinda commented:
> I did mention that:

I was working with the E-Mail I got from GitHub and didn't notice your edit :)

ElementW commented on docs/markdown/Rewriter.md:
```diff
@@ -77,6 +77,17 @@ and print an error message.
 
+### Adding and removing `extra_files`
+
+*Since 0.61.0*
```
Not sure about the version I should specify here, if I should specify at all.

mensinda replied:
Always use the next Meson release version, so this is correct. If a PR doesn't make it the version must be updated.

ElementW commented:
Added unit tests, documentation, as well as release snipped.
The contribution guide mentions

> Must be registered with the FeatureChecks framework

but I don't think this is applicable to the feature I'm adding as this is not a feature usable in build files.

mensinda commented:
> but I don't think this is applicable to the feature I'm adding as this is not a feature usable in build files.

Yes, this is correct because it is a command-line only change.

ElementW commented:
Fixed bits of english and renamed `rewriter_extra_sources.md` to `rewriter_extra_files.md`

ElementW commented:
Fixed the `extra_files argument must be a list` error being triggered on targets with `extra_files` being a variable whose value flattens to an empty array. Added to the `test_target_add_extra_files` test case.

ElementW commented:
Fixed the project language in the test case Meson script, it should resolve the "*Could not get a dynamic linker for build target*" issue on msys2 targets

mensinda reviewed (approved):
LGTM
